### PR TITLE
Allow client to suggest delay until next retry

### PR DESCRIPTION
**Pull request by ast2023**
## What was changed
A new error type was introduced, which allows requesting next retry interval from an activity.

## Why?
Allow developers to adjust retry interval on specific errors.  

## Checklist
1. Closes 1858

2. How was this tested:
Created a workflow/activity. Activity would return the new error with the retry interval. Verified that the value was received by the server.

3. Any docs updates needed?
The new error should be documented in the developer's guide.

### CODE REVIEW

File: temporal/error.go
Comment (ast2023) [inline temporal/error.go:184]: This logic is similar to on in NewApplicationError. Should not we receive the error from the user on NewApplicationErrorWithRetryDelay call?

File: internal/error.go
Comment (bergundy) [inline internal/error.go:132]: I wouldn't put the retry on an error class if we're not serializing it as part of ApplicationFailure.
I still think we want to add the field in ApplicationFailureInfo even if other disagree.

If we must go with the alternative approach, you would probably want to export a `SetNextRetryDelayOnFailure` method on the `activity` package or maybe something more general that could set other attributes as well (although, even though other seem to think so, I disagree that we need to support setting other activity attributes **on failure**).

Comment (Quinn-With-Two-Ns) [reply internal/error.go:132]: +1 to `SetNextRetryDelayOnFailure` method on the activity package

Comment (ast2023) [reply internal/error.go:132]: @bergundy thank you for the recommendation.

> I wouldn't put the retry on an error class if we're not serializing it as part of ApplicationFailure.
> I still think we want to add the field in ApplicationFailureInfo even if other disagree.

ApplicationErrorWithRetryDelay is not an error class. You cannot restore one from a failure. ApplicationErrorWithRetryDelay is an entity which implements Error interface and Delay interface.
You see what I did, I labeled a tail as a leg and now claim that a dog has 5 legs 😄 
But seriously, please help me better understand the reason you think we should not put it on the error.
Do you mean that activity to SDK protocol should define SDK to the server protocol?

> If we must go with the alternative approach, you would probably want to export a SetNextRetryDelayOnFailure method on the activity package or maybe something more general that could set other attributes as well (although, even though other seem to think so, I disagree that we need to support setting other activity attributes on failure).

I let user return 2 things which are related in one call: ApplicationError and RetryDelay. I agree there is some issue that what the call produces is of type error, but I did not want to change the type of what Executor expects for backward computability. Regarding other attributes, they may be set by the user in the way which SDK team finds appropriate. They do not have to follow this if the attribute does not relate to the failure.

Comment (cretz) [reply internal/error.go:132]: > But seriously, please help me better understand the reason you think we should not put it on the error.

IMO the same reason as was suggested in the discussions on this topic that it should not be on the error in the protos - it should not be a product of the error but a settable property anywhere in the activity for when an error occurs.

+1 to SetNextRetryDelayOnFailure method on the activity package

Comment (Sushisource) [reply internal/error.go:132]: As mentioned when we met - I think it can actually go both places and that makes good sense. I think it makes sense to have an API to attach a next delay to any error you're returning, and also the general `SetNextRetryDelayOnFailure` API. Both can be quite useful depending on the structure of your code, and both can end up in the same place in the actual API response.

Comment (Quinn-With-Two-Ns) [reply internal/error.go:132]: If we do want to attach it to the error I think we should have something more general then forcing users to use `ApplicationErrorWithRetryDelay`. I would suggest we check an interface like `retryDelay()` and if the error implements that interface we use it. If we allow both I am not sure which takes precedent.

Comment (Sushisource) [reply internal/error.go:132]: +1 - in my mind you would use an interface just like you've said, which you could augment any error with something that wraps the original and implements that interface. As for precedence, I imagine the per-error one would take precedence over the general "next delay" call, since the former is more specific/granular than the latter.

Comment (cretz) [reply internal/error.go:132]: It was a mistake by the original developers to make `temporal.NewApplicationError` return `error` instead of `*ApplicationError` and thus we can't really attach things easily to it. Luckily our other SDKs don't suffer from this. I am not a fan of having two ways personally, but I don't have strong objection so long as the easier, simpler, clearer way everyone agrees (`activity.SetNextRetryDelayOnFailure`) exists.

I don't suspect there will be many users writing errors conforming to that interface. Looking at the failure converter, using the interface approach, how can I set error type, some details, and next retry delay in my error? Can I see a snippet of code of how that might look?

Comment (Quinn-With-Two-Ns) [reply internal/error.go:132]: >Looking at the failure converter, using the interface approach

The failure converter is not relevant. If you look at the current `extractActivityRequestsFrom` it is called on the error returned from the activity. I am suggesting that function should check for an interface rather then a specific type. I don't want to give our types any special behaviour that users types can't have.

Comment (cretz) [reply internal/error.go:132]: Can you give an example of returning an error from an activity with this new API with error type, some details, and next retry delay set on the application error? That way I can at least visualize what is meant with the interface approach.

File: internal/error.go
Comment (cretz) [inline internal/error.go:141]: Unused type

File: internal/error.go
Comment (cretz) [inline internal/error.go:325]: To match other parts of the Go SDK, can this be:

```go
type ApplicationErrorOptions struct {
    Details []any
    Cause error
    NonRetryable bool
    NextAttemptRetryDelay time.Duration
}

func NewApplicationErrorWithOptions(message, errType string, options ApplicationErrorOptions) error
```

File: internal/failure_converter.go
Comment (cretz) [inline internal/failure_converter.go:106]: Commented out code

File: temporal/error.go
Comment (cretz) [inline temporal/error.go:173]: I don't think there's any value in this being an interface

File: activity/activity.go
Comment (cretz) [inline activity/activity.go:45]: While we don't have to, should probably wait on the name of the proto from https://github.com/temporalio/api/pull/334 and name this similarly. I'm thinking:

```suggestion
	ExtraAttributes = internal.ActivityExtraAttributes
```

But we'll see what that message gets named in that PR. I have an outstanding comment at https://github.com/temporalio/api/pull/334/files#r1422999174. I don't think it's clear to call it `ActivityRequests` there, and no reason not to use similar naming there and here since this will be applied across SDKs as a new Temporal term.

File: activity/activity.go
Comment (cretz) [inline activity/activity.go:117]: Recommend:

```go
func GetExtraAttributes(ctx context.Context) ExtraAttributes {
    return internal.GetActivityExtraAttributes(ctx)
}

func SetExtraAttributes(ctx context.Context, attrs ExtraAttributes) error {
    return internal.SetActivityExtraAttributes(ctx, attrs)
}
```

There are some alternative designs, but I feel this is simplest. Also, please add Godoc to all public API. Also, please mark this as experimental for now until we are confident in the server-side implementation.

File: internal/extra_information.go
Comment (cretz) [inline internal/extra_information.go:1]: Not deserving of a new file IMO

File: temporal/error.go
Comment (Sushisource) [reply temporal/error.go:173]: +1

Comment (bergundy) [reply temporal/error.go:173]: I think we want Chad's way of constructing application error, don't see a reason for including anything in this file other than exporting the internal constructor.

File: internal/error.go
Comment (bergundy) [reply internal/error.go:325]: I would also return `*ApplicationError` here instead of the `error` interface.

File: activity/activity.go
Comment (cretz) [inline activity/activity.go:47]: "extra information" implies that it's something like the existing "GetInfo" (also, it's confusing to call it "information" here when we use "info" elsewhere). I think "extra attributes" is probably best. But we can wait until we see what happens in https://github.com/temporalio/api/pull/334/files#r1422999174. There are a lot of different words like "extra", "information", "requests", etc being used.

File: internal/activity.go
Comment (cretz) [inline internal/activity.go:65]:
```suggestion
		RetryDelay time.Duration
```

Can we just make this a public field like we do with activity info? Also, would prefer if we could avoid pointers to primitives like this and instead just assume unset/0 means unset.

File: activity/activity.go
Comment (cretz) [inline activity/activity.go:119]: I don't think this call is necessary

File: internal/activity.go
Comment (cretz) [inline internal/activity.go:361]: I don't think this call is necessary

File: internal/activity.go
Comment (cretz) [inline internal/activity.go:366]: Would recommend just letting users set this as a field

File: internal/error.go
Comment (cretz) [reply internal/error.go:138]: I don't think you need "extra information" broken out, just put the next retry field on here explicitly for now (though of course I disagree with it being on error at all).

File: temporal/error.go
Comment (cretz) [inline temporal/error.go:241]: I think this should look like https://github.com/temporalio/sdk-go/pull/1319#discussion_r1425330300:

```
type ApplicationErrorOptions struct {
    Details []any
    Cause error
    NonRetryable bool
    NextAttemptRetryDelay time.Duration
}

func NewApplicationErrorWithOptions(message, errType string, options ApplicationErrorOptions) error
```

You don't need all the extra indirection.

File: internal/activity.go
Comment (bergundy) [reply internal/activity.go:366]: Yes please, we don't use the builder pattern in this SDK AFAIK.

File: temporal/error.go
Comment (bergundy) [reply temporal/error.go:241]: Yes please. This is way more inline with the other SDK APIs.

File: temporal/error.go
Comment (cretz) [inline temporal/error.go:181]: I thought per https://github.com/temporalio/sdk-go/pull/1319#discussion_r1425865743 we were going to do something like this:

```go
type ApplicationErrorOptions struct {
    Details []any
    Cause error
    NonRetryable bool
    NextAttemptRetryDelay time.Duration
}

func NewApplicationErrorWithOptions(message, errType string, options ApplicationErrorOptions) error
```

I strongly suggest we take this approach. (will save review of the other parts until after this fundamental part is decided)

Comment (ast2023) [reply temporal/error.go:181]: @cretz This is what I thought we discussed in December. I captured the screenshot for my notes as outcome. The comment you mentioned is regarding moving message and error type out of the error attribvutes structure which makes them required at the compile time. Please let me know if you would like to discuss further. 
Screenshot 2024-01-11 at 2 31 26 PM

Comment (cretz) [reply temporal/error.go:181]: I do think the single `WithOptions` and a single options set is simpler. Especially since now this will only be an error option, this is no different than any other optional field on application error I think.

Comment (bergundy) [reply temporal/error.go:181]: I think it's okay to go with the first option (single Options struct) and document that _currently_ the retry delay is only applicable in activities.

File: temporal/error.go
Comment (Quinn-With-Two-Ns) [inline temporal/error.go:178]: This name should be `NewApplicationErrorWithOptions`

File: internal/error.go
Comment (Quinn-With-Two-Ns) [inline internal/error.go:122]: Per @cretz comments I think this should be `ApplicationErrorOptions` no?

File: internal/extra_requests.go
Comment (Quinn-With-Two-Ns) [inline internal/extra_requests.go:27]: What is this interface for?

File: internal/error.go
Comment (Quinn-With-Two-Ns) [inline internal/error.go:122]: This struct and fields should be documented.

File: temporal/error.go
Comment (Quinn-With-Two-Ns) [inline temporal/error.go:175]: > This function also receives some extra requests

Is this still relevant?

File: temporal/error.go
Comment (cretz) [inline temporal/error.go:176]:
```suggestion
func NewApplicationErrorWithOptions(msg, errType string, options ApplicationErrorOptions) error {
```

File: internal/error.go
Comment (cretz) [inline internal/error.go:314]: Might consider making `NewApplicationError` call this overload internally

File: internal/internal_task_pollers.go
Comment (cretz) [inline internal/internal_task_pollers.go:1]: I don't think any changes need to be made to this file

File: internal/error.go
Comment (cretz) [inline internal/error.go:137]: I think we should add a `func (*ApplicationError) NextRetryDelay() time.Duration` akin to the other getters on this struct

**Review by cretz — COMMENTED** [review]
Looking good, I love the simplicity. Some minor things.

File: internal/error.go
Comment (bergundy) [inline internal/error.go:320]: Can this return `*ApplicationError` and follow the recommended Go practice?

Comment (ast2023) [reply internal/error.go:320]: Could you kindly share the practices you have in mind?

File: internal/error.go
Comment (ast2023) [reply internal/error.go:143]: @cretz these getter functions are only used in tests. When we will need access to the ApplicationError fields outside of the package we can allways add the function.

File: internal/error.go
Comment (ast2023) [reply internal/error.go:128]: @Quinn-With-Two-Ns done

File: internal/error.go
Comment (cretz) [reply internal/error.go]: It is often better Go practice to return the concrete structure (w/ pointer as needed, such as here) instead of the interface (`error`). But we haven't followed that in `NewApplicationError` or `NewApplicationErrorWithCause` or any other `NewXError` call in the `temporal` package, so I can understand both arguments. I do not have an opinion and could totally understand leaving it as is for consistency.

File: internal/error.go
Comment (Quinn-With-Two-Ns) [reply internal/error.go:143]: Users may also use those getters

File: internal/error.go
Comment (Quinn-With-Two-Ns) [inline internal/error.go:126]: Nit:  Generally in the Go SDK we would provide documentation for members on the member variables themselves

File: internal/error.go
Comment (Quinn-With-Two-Ns) [inline internal/error.go:127]: If users do want immediate retries is the "work around" to just pass a very short timeout?

File: internal/error.go
Comment (cretz) [reply internal/error.go:143]: @ast2023 - they are used by users too. Just because they are internal, if they are on the struct they are available.

File: internal/error.go
Comment (Quinn-With-Two-Ns) [inline internal/error.go:128]: Should also note this does not apply to local activities,

File: internal/error.go
Comment (ast2023) [reply internal/error.go:127]: @Quinn-With-Two-Ns you are correct. It is so trivial solution that I did not think it is required to write it down. I can add it.

Comment (Quinn-With-Two-Ns) [reply internal/error.go]: You are correct it is trivial, but sometimes it is nice to point it out anyway for users so they know it is intended. Not a big deal if you don't want to include it.

**Review by Quinn-With-Two-Ns — APPROVED** [review]
LGTM, I'll wait for @cretz or @bergundy to review as well before merging

File: internal/error.go
Comment (cretz) [inline internal/error.go:551]: Should have godoc, but not that big of a deal.

File: internal/failure_converter.go
Comment (cretz) [inline internal/failure_converter.go:101]: Sorry I missed this before. We should leave this as `nil` if `nextRetryDelay` is `0`.

**Review by cretz — COMMENTED** [review]
Sorry, only one minor thing about proto nil and I can mark approved

File: internal/error.go
Comment (ast2023) [reply internal/error.go:551]: I am failing to find documentation which will explain NextRetryDelay better than the name of the method.

Comment (cretz) [reply internal/error.go:551]: Godoc rules often stipulate Godoc for all exported items regardless of obviousness (yes that means usually if redundant). Granted we don't have such a strict requirement, but we usually follow it.

**Pull request merged**
